### PR TITLE
utility track data

### PR DESCRIPTION
**Pull request by mfdeveloper508**

### CODE REVIEW

**Comment by mfdeveloper508** [comment]
Track 3 has several elements in security data and additional data
but track 3 has been implemented easily using dummy element
Should we implement track 3 with detailed data format?

**Comment by wadearnold** [comment]
@mfdeveloper508  let's pull out the track 3 secure element data into its own issue and address it separately. What you have implemented will work for a magnetic card swipe.

**Review by wadearnold — APPROVED** [review]
@alovak can you merge this after your review and cut a release with a version and release notes.

File: track/regen.go
Comment (krishishah) [inline track/regen.go:34]: Any reason we're publicly exposing this func as well as the `Generator` interface? I don't

**Comment by alovak** [comment]
I'm on it.

**Comment by alovak** [comment]
@mfdeveloper508 thanks for a great job! I see 

I want to suggest some changes before we merge it with the master.

It would be great to implement track1, track2 (and maybe track3) as a `field.Field`, not just as a utility class. If track implements the `field.Field` interface, we can easily use it in our specifications. We can also use them separately as utility class. Here is what I mean:

```go
// Example 1, create track 1 field using
track1 := track.NewTrack1Value(&track.Track1Data{
	Number:      "",
	Name:        "",
	ServiceCode: "",
	// ...
})

raw, err := track1.Pack()
// handle error

track1 := field.NewTrack1(&field.Spec{
	Length:      76,
	Description: "Track 1 Data",
	Enc:         encoding.ASCII,
	Pref:        prefix.ASCII.LL,
})

read, err := track1.Unpack("011234567890123445=724724000000000****00300XXXX020200099010=********************==1=100000000000000000**")
// handle error

trackData := track1.Data().(*track.Track1Data)
trackData.ServiceCode
trackData.Number
trackData.Name
trackData.CVV
trackData.DiscretionaryData
// ....

```

Changes that are required:
* move `./track` into `./field` -> `/field/track/track1`
* implement `Field` interface for `Track1` and `Track2`
* add `Data()` method so we can get value out of the track field

Thoughts, comments?

File: track/track.go
Comment (alovak) [inline track/track.go:42]: While I see the benefit of detecting card type, I think that current model should be changed a little bit into something like:

* Card
  * Track 1
  * Track 2

It means that `Card` contains tracks and then we can add `Type` into `Card` object.

If there is no specific need for CardType here, let's remove it.

File: track/track.go
Comment (alovak) [inline track/track.go:59]: As I suggested in my first comment to the PR, let's implement `field.Field` interface in `Track1` and `Track2`.

File: track/track.go
Comment (alovak) [inline track/track.go:117]: I would use `regexp.MustCompile(...)` at the package level (outside of the function) to compile once and just use compiled regexp in the function then.

**Review by krishishah — COMMENTED** [review]
I agree with @alovak 's comments on the whole. It feels more intuitive to implement much of this within the `field` package.

I'd also rename `First`, `Second`, `Third` as `Track1`, `Track2`, `Track3` etc

I personally don't see the value of the `Generator` pattern myself and see lots of opportunity to simplify the code in its current state. However, it's possible that I don't fully appreciate what its there for.

As a general pointer, could we also document all of our public structs, fields, functions and methods and add examples wherever possible? It makes it much easier to follow the code.

File: track/track.go
Comment (krishishah) [reply track/track.go:34]: If we're moving this to be within `field/track` then we could even just call it `Data`?

No need to repeat `track.TrackData`.

But :100: to the rename from `GeneralCard` to something more concise

**Comment by krishishah** [comment]
> move ./track into ./field -> /field/track/track1

@alovak is there a specific reason you don't just want to have this logic live inside of `field/`? Do we need the nested `field/track` pkg?

File: track/track.go
Comment (alovak) [inline track/track.go:137]: should it include `/`?

File: track/track.go
Comment (alovak) [inline track/track.go:105]: Not sure I see the benefit from introduced `ExpiryDate` type. If we use `time.Time` in `GeneralCard` (or in `TrackData` when we change it) we just need two functions:
1. to parse/return `time.Time`
2. to return string representation like `t.Format("06/01")`

File: track/track_test.go
Comment (alovak) [inline track/track_test.go:20]: It would be great if we test parsed value here as well. From my experience, it's not enough to read/write without error. We should check what exactly we get in `Number`, `Expiration` and other fields.

File: track/track_test.go
Comment (alovak) [inline track/track_test.go:34]: From the name of the test, I would expect us to test for some error here. But I don't see it. What's the expected behavior? Not to fail (ignore name length?)?

File: track/track_test.go
Comment (alovak) [inline track/track_test.go:52]: As in my previous comment, we need to check if we parse values properly. As with table tests I would use `input` and `want` structure with expected values.

**Comment by alovak** [comment]
> > move ./track into ./field -> /field/track/track1
> 
> @alovak is there a specific reason you don't just want to have this logic live inside of `field/`? Do we need the nested `field/track` pkg?

We can put it just into field. Initially I thought that we will need more files to support `track` field. But it seems we can put it all into one file.

File: track/regen.go
Comment (alovak) [inline track/regen.go:60]: Can we do the same without `init()`?

File: track/track.go
Comment (alovak) [reply track/track.go:34]: It seems we don't even need `Data` as `Track1` is already that field. At least this is how it works for `String`, `Number` and the rest.

File: track/regen.go
Comment (mfdeveloper508) [reply track/regen.go:1]: that's right, 
'reqen' is short for regenerator of regexp
main feature is to create string with regex pattern and params

File: field/track.go
Comment (louisheath) [inline field/track.go:59]: If these are to be of type `TrackVersion` lets define them so, as currently they are `int`s
```suggestion
type TrackNumber int

const (
    Track1 TrackNumber = 1
    Track2 TrackNumber = 2
    Track3 TrackNumber = 3
)
```
Within this suggestion I also renamed the variables as `field.VersionFirst` to me doesn't sound related to track data.
What do you think?

File: field/track.go
Comment (louisheath) [inline field/track.go:69]: To avoid possible copy paste errors here we could consider using constants
```suggestion
const (
	expiryDateFormat   = "0601"
	
	track1Pattern = `^([A-Z]{1})([0-9]{1,19})\^([^\^]{2,26})\^([0-9]{4}|\^)([0-9]{3}|\^)([^\?]+)$`
	track2Pattern = `^([0-9]{1,19})\=([0-9]{4}|\=)([0-9]{3}|\=)([^\?]+)$`
	track3Pattern = `^([0-9]{2})([0-9]{1,19})\=([^\?]+)$`
)

var (
	track1Regex = regexp.MustCompile(track1Pattern)
	track2Regex = regexp.MustCompile(track2Pattern)
	track3Regex = regexp.MustCompile(track3Pattern)
)
```

File: field/track.go
Comment (louisheath) [inline field/track.go:84]: Do we need `version` to be a parameter if this function is specific to track 1?

File: field/track.go
Comment (louisheath) [inline field/track.go:81]: nit: I wasn't sure what `isFixed` meant at first
 ```suggestion
func NewTrack1Value(val []byte, version TrackVersion, fixedLength bool) (*Track, error) {
```

File: field/track.go
Comment (louisheath) [inline field/track.go:395]: So I'm understanding `regen` to be a way of generating a random string using a regular expression,
and then using these handlers to inject strings we actually want at each index. This feels long winded.

Would something like `fmt.Sprintf("%s%s%s", a, b, c)` not suffice?

File: utils/regen.go
Comment (louisheath) [inline utils/regen.go:3]: I think this file warrants more documentation to help readers understand its purpose

**Review by louisheath — COMMENTED** [review]
Thanks @mfdeveloper508, this PR is a massive help for us over at TM. Really excited to see this in master

(Apologies for the mass of separate comments, I haven't reviewed on Github in a while!)

File: field/track.go
Comment (mfdeveloper508) [reply field/track.go:207]: the issue is originated to adding utility functions for track
so I am trying to add some stuffs for track
I hope that the utility function will be use to check card type

**Comment by krishishah** [comment]
As of https://github.com/moov-io/iso8583/pull/127/files, fields now should implement the `json.Unmarshaler` interface and `SetBytes()` now also sets bytes for the original data struct if one already exists. The changes to `field/composite.go` should give you a decent idea of what's involved there.

Could you update your PR to reflect this please? :pray:

**Comment by mfdeveloper508** [comment]
> As of https://github.com/moov-io/iso8583/pull/127/files, fields now should implement the `json.Unmarshaler` interface and `SetBytes()` now also sets bytes for the original data struct if one already exists. The changes to `field/composite.go` should give you a decent idea of what's involved there.
> 
> Could you update your PR to reflect this please? 🙏

**Comment by mfdeveloper508** [comment]
> > As of https://github.com/moov-io/iso8583/pull/127/files, fields now should implement the `json.Unmarshaler` interface and `SetBytes()` now also sets bytes for the original data struct if one already exists. The changes to `field/composite.go` should give you a decent idea of what's involved there.
> > Could you update your PR to reflect this please? 🙏

I see, will check again

File: field/track.go
Comment (alovak) [inline field/track.go:53]: @mfdeveloper508 can we split Track into separate structs/fields: Track1, Track2 and Track3? It seems that it will simplify the code and make it easier to understand and maintain.

File: field/track.go
Comment (alovak) [inline field/track.go:14]: why do we need `xml` tags?

**Comment by wadearnold** [comment]
Hey, @alovak @krishishah @louisheath these are pretty small refactors. I agree with everything suggested. @mfdeveloper508 is grinding down some mission-critical issues for ISO20022. If we want to get this merged can some contribute some of these reactors so we can get this merged?

**Comment by krishishah** [comment]
> Hey, @alovak @krishishah @louisheath these are pretty small refactors. I agree with everything suggested. @mfdeveloper508 is grinding down some mission-critical issues for ISO20022. If we want to get this merged can some contribute some of these reactors so we can get this merged?

I don't think the current implementation works for the typed API (which we use in our production setup) to pack, unpack etc. However, I understand that the PR is dragging on a bit. I'll take a step back from this PR and push a follow up to plug any gaps we see in the implementation.

Thanks!

File: field/track/track1.go
Comment (alovak) [inline field/track/track1.go:30]:
```suggestion
```

File: field/track/track1.go
Comment (alovak) [inline field/track/track1.go:29]: let's move it to the related files

File: field/track/track1.go
Comment (alovak) [inline field/track/track1.go:35]: let's move it to the related files

**Comment by alovak** [comment]
@krishishah I've updated this PR and now it supports typed API as well. Also, tracks are now under the `field` package.

**Comment by alovak** [comment]
I'm not sure about `NewTrackXValue` functions... I just removed them for a now as I don't see what is the value in them.

**Pull request merged**
